### PR TITLE
Add footnote label displayed at the bottom of an instruction or form step.

### PR DESCRIPTION
**Pull request by syoung-smallwisdom**
This is useful for displaying disclaimer or copyright information.

simulator screen shot oct 13 2016 12 35 27 pm

### CODE REVIEW

**Comment by umerkhan-apple** [comment]
@syoung-smallwisdom Could you resolve the merge conflicts? Thanks!

**Comment by syoung-smallwisdom** [comment]
@umerkhan-apple 
Conflicts resolved.

**Comment by umerkhan-apple** [comment]
@syoung-smallwisdom Thank you. Pending internal review from design team.

File: ResearchKit/Common/ORKFootnoteLabel.m [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "ORKFootnoteLabel.h"
+#import "ORKSkin.h"
+
```
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKFootnoteLabel.m]: 2 new lines before and after imports

File: ResearchKit/Common/ORKFormStep.m [hunk 2]
```diff
@@ -92,31 +92,36 @@ - (void)validateIdentifiersUnique {
 - (instancetype)copyWithZone:(NSZone *)zone {
     ORKFormStep *step = [super copyWithZone:zone];
     step.formItems = ORKArrayCopyObjects(_formItems);
+    step.footnote = self.footnote;
```
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKFormStep.m]: [self.footnote copy]

File: ResearchKit/Common/ORKFormStep.m [hunk 3]
```diff
@@ -92,31 +92,36 @@ - (void)validateIdentifiersUnique {
...
 - (NSUInteger)hash {
-    return super.hash ^ self.formItems.hash;
+    return super.hash ^ self.formItems.hash ^ self.footnote.hash;
 }
 
 - (instancetype)initWithCoder:(NSCoder *)aDecoder {
     self = [super initWithCoder:aDecoder];
     if (self) {
         ORK_DECODE_OBJ_ARRAY(aDecoder, formItems, ORKFormItem);
+        ORK_DECODE_OBJ(aDecoder, footnote);
```
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKFormStep.m]: ORK_DECODE_OBJ_CLASS(aDecoder, footnote, NSString)

File: ResearchKit/Common/ORKInstructionStep.m [hunk 4]
```diff
@@ -77,6 +79,7 @@ + (BOOL)supportsSecureCoding {
 - (instancetype)copyWithZone:(NSZone *)zone {
     ORKInstructionStep *step = [super copyWithZone:zone];
     step.detailText = self.detailText;
+    step.footnote = self.footnote;
```
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKInstructionStep.m]: = [self.footnote copy]

File: ResearchKit/Common/ORKInstructionStep.m [hunk 5]
```diff
@@ -77,6 +79,7 @@ + (BOOL)supportsSecureCoding {
 - (instancetype)copyWithZone:(NSZone *)zone {
     ORKInstructionStep *step = [super copyWithZone:zone];
     step.detailText = self.detailText;
```
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKInstructionStep.m]: Not part of your changes, but can you also update this to be [self.detailText copy]

File: ResearchKit/Common/ORKInstructionStepView.m [hunk 6]
```diff
@@ -143,6 +144,7 @@ - (void)setInstructionStep:(ORKInstructionStep *)instructionStep {
     self.headerView.iconImageView.image = _instructionStep.iconImage;
     self.headerView.captionLabel.text = _instructionStep.title;
     
+
```
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKInstructionStepView.m]: remove extra line

File: ResearchKit/Common/ORKNavigationContainerView_Internal.h [hunk 7]
```diff
@@ -32,14 +32,15 @@
 #import "ORKNavigationContainerView.h"
 #import "ORKContinueButton.h"
 #import "ORKTextButton_Internal.h"
-
+#import "ORKFootnoteLabel.h"
```
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKNavigationContainerView_Internal.h]: add extra new line.

File: ResearchKit/Common/ORKFormStep.m (hunk 2 above)
Comment (syoung-smallwisdom) [inline ResearchKit/Common/ORKFormStep.m]:
````
@property (nonatomic, copy, nullable) NSString *footnote;
````

Already does this. What you are suggesting would do the following:
````
step->_footnote = [[self.footnote copy] copy];
````

This is different from the `ORKArrayCopyObjects` macro which copies the objects in the array vs. `step.formItems = self.formItems` which just copies the pointers to the existing objects.

File: ResearchKit/Common/ORKInstructionStep.m (hunk 5 above)
Comment (syoung-smallwisdom) [inline ResearchKit/Common/ORKInstructionStep.m]: `detailText` is also a copy property.

File: ResearchKit/Common/ORKInstructionStep.m (hunk 4 above)
Comment (syoung-smallwisdom) [inline ResearchKit/Common/ORKInstructionStep.m]: copy property

**Comment by syoung-smallwisdom** [comment]
@umerkhan-apple Updated or replied. Thanks!

File: ResearchKit/Common/ORKFormStep.m (hunk 2 above)
Comment (umerkhan-apple) [inline ResearchKit/Common/ORKFormStep.m]: Ah thats right! Sounds good. PR looks good otherwise!

**Pull request merged**